### ISSUE TITLE
Help: Queues can not be deserialized on iOS devices

### ISSUE THREAD

**Issue by felipegodias** [jilleJr/Newtonsoft.Json-for-Unity#51]
##  Expected behavior
Queue<T> is successfully deserialized on iOS devices.

## Actual behavior
JsonSerializationException is thrown when deserializing generic Queues.

Serialization works fine in all environments.
Deserialization works on Mac Editor, Windows Editor and Android builds.

```
2020-05-06 12:38:53.867366-0300 dev[16792:5031829] Built from '2019.3/staging' branch, Version '2019.3.10f1 (5968d7f82152)', Build type 'Release', Scripting Backend 'il2cpp'
-> applicationDidFinishLaunching()
-> applicationDidBecomeActive()
GfxDevice: creating device client; threaded=1
Initializing Metal device caps: Apple A11 GPU
Initialize engine version: 2019.3.10f1 (5968d7f82152)
2020-05-06 12:38:54.415064-0300 dev[16792:5031829] Unbalanced calls to begin/end appearance transitions for <SplashScreenController: 0x10847c4a0>.
UnloadTime: 0.187333 ms
{"$type":"System.Collections.Generic.Queue`1[[System.Int32, mscorlib]], System","$values":[0]}
Test:TestQueue(JsonSerializerSettings)
Test:Start()
 
(Filename: ./Runtime/Export/Debug/Debug.bindings.h Line: 35)

JsonSerializationException: Cannot create and populate list type System.Collections.Generic.Queue`1[System.Int32]. Path '$values', line 1, position 91.
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewList (Newtonsoft.Json.JsonReader reader, Newtonsoft.Json.Serialization.JsonArrayContract contract, System.Boolean& createdFromNonDefaultCreator) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateList (Newtonsoft.Json.JsonReader reader, System.Type objectType, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonProperty member, System.Object existingValue, System.String id) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ReadMetadataProperties (Newtonsoft.Json.JsonReader reader, System.Type& objectType, Newtonsoft.Json.Serialization.JsonContract& contract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerMember, System.Object existingValue, System.Object& newValue, System.String& id) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject (Newtonsoft.Json.JsonReader reader, System.Type objectType, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerMember, System.Object existingValue) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal (Newtonsoft.Json.JsonReader reader, System.Type objectType, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerMember, System.Object existingValue) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize (Newtonsoft.Json.JsonReader reader, System.Type objectType, System.Boolean checkAdditionalContent) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.JsonSerializer.DeserializeInternal (Newtonsoft.Json.JsonReader reader, System.Type objectType) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.JsonConvert.DeserializeObject (System.String value, System.Type type, Newtonsoft.Json.JsonSerializerSettings settings) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.JsonConvert.DeserializeObject[T] (System.String value, Newtonsoft.Json.JsonSerializerSettings settings) [0x00000] in <00000000000000000000000000000000>:0 
  at Test.TestQueue[T] (Newtonsoft.Json.JsonSerializerSettings settings) [0x00000] in <00000000000000000000000000000000>:0 

  at Test.Start () [0x00000] in <00000000000000000000000000000000>:0 
Test:Start()
 
(Filename: currently not available on il2cpp Line: -1)

{"$type":"System.Collections.Generic.Queue`1[[System.String, mscorlib]], System","$values":[null]}
Test:TestQueue(JsonSerializerSettings)
 
(Filename: ./Runtime/Export/Debug/Debug.bindings.h Line: 35)

JsonSerializationException: Cannot create and populate list type System.Collections.Generic.Queue`1[System.String]. Path '$values', line 1, position 92.
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewList (Newtonsoft.Json.JsonReader reader, Newtonsoft.Json.Serialization.JsonArrayContract contract, System.Boolean& createdFromNonDefaultCreator) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateList (Newtonsoft.Json.JsonReader reader, System.Type objectType, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonProperty member, System.Object existingValue, System.String id) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ReadMetadataProperties (Newtonsoft.Json.JsonReader reader, System.Type& objectType, Newtonsoft.Json.Serialization.JsonContract& contract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerMember, System.Object existingValue, System.Object& newValue, System.String& id) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject (Newtonsoft.Json.JsonReader reader, System.Type objectType, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerMember, System.Object existingValue) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal (Newtonsoft.Json.JsonReader reader, System.Type objectType, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerMember, System.Object existingValue) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize (Newtonsoft.Json.JsonReader reader, System.Type objectType, System.Boolean checkAdditionalContent) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.JsonSerializer.DeserializeInternal (Newtonsoft.Json.JsonReader reader, System.Type objectType) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.JsonConvert.DeserializeObject (System.String value, System.Type type, Newtonsoft.Json.JsonSerializerSettings settings) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.JsonConvert.DeserializeObject[T] (System.String value, Newtonsoft.Json.JsonSerializerSettings settings) [0x00000] in <00000000000000000000000000000000>:0 
  at Test.TestQueue[T] (Newtonsoft.Json.JsonSerializerSettings settings) [0x00000] in <00000000000000000000000000000000>:0 
 
(Filename: currently not available on il2cpp Line: -1)
```

## Steps to reproduce

- New project
- Import `jillejr.newtonsoft.json-for-unity` via UPM
- Add following script to scene:

```csharp
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Utilities;
using UnityEngine;

public class Test : MonoBehaviour
{
    private static void TestQueue<T>(JsonSerializerSettings settings)
    {
        AotHelper.EnsureList<T>();

        var queue = new Queue<T>();
        queue.Enqueue(default);
        var queueJson = JsonConvert.SerializeObject(queue, settings);
        Debug.Log(queueJson);

        try
        {
            JsonConvert.DeserializeObject<Queue<T>>(queueJson, settings);
            Debug.Log($"JsonConvert.DeserializeObject<Queue<{typeof(T).Name}>>(queueJson, settings);");
        }
        catch (Exception e)
        {
            Debug.LogException(e);
        }
    }

    private void Start()
    {
        var settings = new JsonSerializerSettings
        {
            TypeNameHandling = TypeNameHandling.All
        };

        TestQueue<int>(settings);
        TestQueue<string>(settings);
    }
}
```
- Build XCode Project
- Build Project on XCode 
- Run the game

## Details
Host machine OS running Unity Editor 👉 macOS Catalina 10.15.4, Windows 10 PRO 1909 
Unity build target 👉 iOS
Newtonsoft.Json-for-Unity package version 👉 12.0.101, 12.0.301
I was using Unity version 👉 2019.3.10f1
## Checklist
- [x] Shutdown Unity, deleted the /Library folder, opened project again in Unity, and problem still remains.
- [X] Checked to be using latest version of the package.

**Comment by jilleJr** [comment]
Hi @felipegodias, thanks for reporting this. This looks like some tasty glitch.

You're both using the EnsureList and explicitly constructing a new Queue type, which are the two easy to spot pitfalls.

I cannot give you a straight up answer by just looking at your example. Some ideas comes to mind, but I'll have to sit down with this to further investigate. Probably within a day or two.

**Comment by jilleJr** [comment]
Very rich issue as well. Kudos

**Comment by jilleJr** [comment]
Ha! Interesting find! I was able to reproduce this when building to Windows with IL2CPP. Here's what's going on:

The `Query<>` type is neither an `IList<>` nor an `ICollection<>`, but instead only a `IReadOnlyCollection<>`, which makes it so it doesn't contain any of the `.Add()` methods that comes with `IList<>` and `ICollection<>`.

How the official Json .NET package handles this is by trying to obtain the parameterized constructor `new Queue<T>(IEnumerable<T>)`, which has been omitted by the bytecode stripping when compiling with IL2CPP and therefore you receive the error that it cannot be deserialized.

See in the source for reference, where declaring if it's deserializable actually lands on if it can find the parameterized constructor or not:

https://github.com/JamesNK/Newtonsoft.Json/blob/12.0.3/Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs#L194-L201

## Workaround (short-term fix)

A quickfix would be to ensure that constructor, for example I was getting it to work as intended with the following:

```cs
AotHelper.EnsureList<T>();
AotHelper.Ensure(() =>
{
    _ = new Queue<T>(Array.Empty<T>());
});
```

Looking at the implementation of the [queue.cs in .NET Framework 4.8](https://referencesource.microsoft.com/#mscorlib/system/collections/queue.cs) there's no reference to any `List<T>`, and Json .NET creates a temporary `List<T>` when deserializing the `Queue<>` type (See <https://github.com/JamesNK/Newtonsoft.Json/blob/12.0.3/Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs#L299-L313>) so it should in theory also need the `AotHelper.EnsureList<T>()` call but through my testing it works fine when not including the `AotHelper.EnsureList<T>()` call. I would take the safe path an also include it just in case, but you could omit the `EnsureList<T>` there.

## Long-term fix

Still brainstorming about this, but I'll either add some special logic to the main Json .NET code to handle `Queue<>` (and perhaps other types such as `Stack<>` and even non-generic `Queue` and `Stack`) better, and then add some method to AotHelper, such as `AotHelper.EnsureQueue<>()` or something.

Gets tricky when trying to consider that some may inherit from `Queue<>` and so ensuring usage of the parameterized constructor may not be a viable option.

Eh, these are just brainstorm ideas anyway. I'll figure something out and deploy a 12.0.302 version. If you have any ideas yourself then please share ;) Like how would you as a user want to deal with this? Fine line of allowing fine grained adjustment such as `EnsureQueue` vs to just ensure queues and stacks and hashsets and whatnot all in `EnsureList`.

---

Closing this issue as resolved as there is a workaround and moving the fix over to #53
